packages/indicators/package.json

```diff
@@ -14,9 +14,8 @@
     "build": "babel src --out-dir dist --ignore \"src/__tests__/**\" --extensions .ts --config-file \"../../.babelrc-ts.js\"",
     "lint": "tsc; eslint . --ext .ts",
     "lint:fix": "yarn lint --fix",
-    "test": "jest",
-    "test:coverage": "jest --coverage",
-    "test:watch": "jest --watch"
+    "test": "jest --runInBand",
```

Comment:
This `runInBand` config basically runs the tests in serial rather than parallel right? You'd recommend we use this wherever the tests use the database?

Reply:
Exactly! This has been giving us some pain since we migrate from mocha to jest. I made the change in packages that cause us trouble for now, but perhaps we should proactively update the others... What do you think?

Another idea:
* Explicitly separate integration (using an actual DB) and unit tests in our code base. Either use different folders, or different file extensions
* Run unit tests in parallel and integration tests serially (for now)
* In the future, enabling parallel integration testing by using a new DB for each testing thread. I'm not familiar with this pattern but I briefly read about it yesterday

I think it's high time we don't use a local DB for running the tests:
1. The DB in our CI/CD starts with a clean state, so the local and CI/CD test dbs do not match
2. We have some elaborate data cleaning mechanism to reset the DB to the same state it was before the tests started running. It would make things simpler to start with a "clean" (or populated with test setup data) DB. It would also make all DB operations in the tests faster

Tagging you too here @edmofro so that you are aware of these findings and thoughts

Reply:
(Relevant comment above: https://github.com/beyondessential/tupaia/pull/2850#discussion_r672276857)

Reply:
This is all so true! ​We should absolutely have a separate test db. I believe McLean even tried to set this up at one point, but ran into enough obstacles that he backed out. But yes, for all of those reasons you say, it'd be ideal to have a separate db spun up for tests, and even better a separate one for each thread!

Note that in order to make tests independent, we'd still need either the same sort of data cleaning protocols, a drop and restore, or tests that only make db changes that other tests can live with (difficult to be certain). So long as you're using the same database across more than one test (e.g. per thread), you need to make sure earlier tests don't interfere with it. 

Final thing - in Tamanu they use SQLite for testing which makes things fast and lightweight, but does lead to a) some additional burden maintaining SQLite support and b) some discrepancies where a test may show a false negative or false positive because there is a slight behaviour difference between SQLite and Postgresql. For those reasons I'd like to keep our test db using postgres.

Could you make a card for this @kael89?

Reply:
I agree with using PostgreSQL in out test db (I also came across this discrepancies), also we should certainly clean up the test DB between test runs! But using a clean DB as a starting point will make this easier (eg you can truncate some tables, which is much faster than deleting selected records)

Card created: https://github.com/beyondessential/tupaia-backlog/issues/3210